File: PCG Engine/level-generation/tests/src/SimpleLevelGeneration.cpp
#include <pcg/engine/level-generation/SimpleGeneration.hpp>

#include <gtest/gtest.h>

namespace pcg::engine::level_generation::tests
{
    struct SimpleLevelGenerationParam
    {
        SimpleLevelGenerationParam(math::Vector3 offset) : offset(offset)
        {
        }

        void setCallback()
        {
            if (offset == math::Vector3::right)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.x, points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::left)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.x, -points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::up)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.y, points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::down)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.y, -points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::forward)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.z, points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::backward)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.z, -points[index]);
                        index += 1;
                    };
            }
            else if (offset == math::Vector3::zero)
            {
                callback = [this](math::Vector3 position)
                    {
                        EXPECT_EQ(position.x, 0);
                        EXPECT_EQ(position.y, 0);
                        EXPECT_EQ(position.z, 0);
                    };
            }
        }

        math::Vector3 offset;
        std::function<void(math::Vector3)> callback = nullptr;

    private:
        std::vector<int> points{ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
        int index = 0;
    };

    class SimpleLevelGenerationTest : public ::testing::TestWithParam<SimpleLevelGenerationParam>
    {
    public:
        struct PrintToStringParamName
        {
            template <class ParamType>
            std::string operator()(const testing::TestParamInfo<ParamType>& info) const
            {
                const auto& offset = static_cast<SimpleLevelGenerationParam>(info.param).offset;

                if (offset == math::Vector3::right)
                {
                    return "PositiveX";
                }
                else if (offset == math::Vector3::left)
                {
                    return "NegativeX";
                }
                else if (offset == math::Vector3::up)
                {
                    return "PositiveY";
                }
                else if (offset == math::Vector3::down)
                {
                    return "NegativeY";
                }
                else if (offset == math::Vector3::forward)
                {
                    return "PositiveZ";
                }
                else if (offset == math::Vector3::backward)
                {
                    return "NegativeZ";
                }
                else if (offset == math::Vector3::zero)
                {
                    return "Zero";
                }
            };
        };
    };

    TEST_P(SimpleLevelGenerationTest, SimpleLevelGeneration)
    {
        GenerationData data{ 10, 1, math::Vector3::zero };
        SimpleLevelGenerationParam levelGenerationParam = GetParam();
        levelGenerationParam.setCallback();

        simpleGeneration(data, levelGenerationParam.offset, levelGenerationParam.callback);
    }

    INSTANTIATE_TEST_CASE_P(
        SimpleLevelGenerationTests,
        SimpleLevelGenerationTest,
        ::testing::Values(
            SimpleLevelGenerationParam(math::Vector3::right),
            SimpleLevelGenerationParam(math::Vector3::left),
            SimpleLevelGenerationParam(math::Vector3::up),
            SimpleLevelGenerationParam(math::Vector3::down),
            SimpleLevelGenerationParam(math::Vector3::forward),
            SimpleLevelGenerationParam(math::Vector3::backward),
            SimpleLevelGenerationParam(math::Vector3::zero)
        ),
        SimpleLevelGenerationTest::PrintToStringParamName());
}

File: C#-PCG-Wrapper/Wrapper-Test/SimpleLevelGeneration.cs
using System.Collections;

namespace PCGAPI.Tests.LevelGeneration
{
    public class SimpleLevelGeneration : PCGTest
    {
        int index = 0;
        readonly int[] points = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

        class SimpleLevelGenerationData : IEnumerable<object[]>
        {
            public static readonly Vector3 right = new(1, 0, 0);
            public static readonly Vector3 left = new(-1, 0, 0);
            public static readonly Vector3 up = new(0, 1, 0);
            public static readonly Vector3 down = new(0, -1, 0);
            public static readonly Vector3 forward = new(0, 0, 1);
            public static readonly Vector3 backward = new(0, 0, -1);
            public static readonly Vector3 zero = new(0, 0, 0);

            public IEnumerator<object[]> GetEnumerator()
            {
                yield return new object[] { right };
                yield return new object[] { left };
                yield return new object[] { up };
                yield return new object[] { down };
                yield return new object[] { forward };
                yield return new object[] { backward };
                yield return new object[] { zero };
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                return GetEnumerator();
            }
        }

        [Theory]
        [ClassData(typeof(SimpleLevelGenerationData))]
        public void SimpleGeneration(Vector3 offset)
        {
            GenerationParameters parameters = new()
            {
                count = 10,
                size = 1,
            };

            PCGEngine.SimpleGeneration(parameters, offset, position =>
            {
                if (offset == SimpleLevelGenerationData.right)
                {
                    Assert.Equal(points[index], position.x);
                }
                if (offset == SimpleLevelGenerationData.left)
                {
                    Assert.Equal(-points[index], position.x);
                }
                if (offset == SimpleLevelGenerationData.up)
                {
                    Assert.Equal(points[index], position.y);
                }
                if (offset == SimpleLevelGenerationData.down)
                {
                    Assert.Equal(-points[index], position.y);
                }
                if (offset == SimpleLevelGenerationData.forward)
                {
                    Assert.Equal(points[index], position.z);
                }
                if (offset == SimpleLevelGenerationData.backward)
                {
                    Assert.Equal(-points[index], position.z);
                }
                if (offset == SimpleLevelGenerationData.zero)
                {
                    Assert.Equal(0, position.x);
                    Assert.Equal(0, position.y);
                    Assert.Equal(0, position.z);
                }

                index += 1;
            });
        }
    }
}
